sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -0,0 +1,228 @@
+    assert A.dcm(B) == Matrix([[cos(q2), -sin(q2), 0], [sin(q2), cos(q2), 0], [0, 0, 1]])
+    assert N.dcm(B).simplify() == Matrix([[cos(q1 + q2), -sin(q1 + q2), 0], \
+        [sin(q1 + q2), cos(q1 + q2), 0], [0, 0, 1]])
+
+    #Check Angular Velocity
+    assert A.ang_vel_in(N) == u1 * N.z
+    assert B.ang_vel_in(A) == u2 * A.z
+    assert B.ang_vel_in(N) == u1 * N.z + u2 * A.z
+
+    #Check kde
+    assert J1.kdes() == [u1 - q1.diff(t)]
+    assert J2.kdes() == [u2 - q2.diff(t)]
+
+    #Check Linear Velocity
+    assert PartP.masscenter.vel(N) == l*u1*A.y
+    assert PartR.masscenter.vel(A) == l*u2*B.y
+
+def test_pin_joint_chaos_pendulum():
+    mA, mB, lA, lB, h = symbols('mA, mB, lA, lB, h')
+    theta, phi, omega, alpha = dynamicsymbols('theta phi omega alpha')
+    N = ReferenceFrame('N')
+    A = ReferenceFrame('A')
+    B = ReferenceFrame('B')
+    lA = (lB - h / 2) / 2
+    lC = (lB/2 + h/4)
+    rod = Body('rod', frame=A, mass=mA)
+    plate = Body('plate', mass=mB, frame=B)
+    C = Body('C', frame=N)
+    J1 = PinJoint('J1', C, rod, coordinates=theta, speeds=omega, child_joint_pos=lA*A.z,\
+        parent_axis=N.y, child_axis=A.y)
+    J2 = PinJoint('J2',rod, plate, coordinates=phi, speeds=alpha, parent_joint_pos=lC*A.z,\
+        parent_axis=A.z, child_axis=B.z)
+
+    #Check orientation
+    assert A.dcm(N) == Matrix([[cos(theta), 0, -sin(theta)], [0, 1, 0], [sin(theta), 0, cos(theta)]])
+    assert A.dcm(B) == Matrix([[cos(phi), -sin(phi), 0], [sin(phi), cos(phi), 0], [0, 0, 1]])
+    assert B.dcm(N) == Matrix([[cos(phi)*cos(theta), sin(phi), -sin(theta)*cos(phi)], \
+        [-sin(phi)*cos(theta), cos(phi), sin(phi)*sin(theta)], [sin(theta), 0, cos(theta)]])
+
+    #Check Angular Velocity
+    assert A.ang_vel_in(N) == omega * N.y
+    assert A.ang_vel_in(B) == - alpha * A.z
+    assert N.ang_vel_in(B) == - omega * N.y - alpha * A.z
+
+    #Check kde
+    assert J1.kdes() == [omega - theta.diff(t)]
+    assert J2.kdes() == [alpha - phi.diff(t)]
+
+    #Check pos of masscenters
+    assert C.masscenter.pos_from(rod.masscenter) == lA*A.z
+    assert rod.masscenter.pos_from(plate.masscenter) == - lC * A.z
+
+    #Check Linear Velocities
+    assert rod.masscenter.vel(N) == (h/4 - lB/2)*omega*A.x
+    assert plate.masscenter.vel(N) == ((h/4 - lB/2)*omega + (h/4 + lB/2)*omega)*A.x
+
+def test_pinjoint_arbitrary_axis():
+    def generate_body():
+        N = ReferenceFrame('N')
+        A = ReferenceFrame('A')
+        P = Body('P', frame=N)
+        C = Body('C', frame=A)
+        return N, A, P, C
+
+    theta, omega = dynamicsymbols('J_theta, J_omega', real=True)
+
+    #When the bodies are attached though masscenters but axess are opposite.
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, child_axis=-A.x)
+
+    assert -A.x.angle_between(N.x) == 0
+    assert -A.x.express(N) == -N.x
+    assert A.dcm(N) == Matrix([[1, 0, 0], [0, cos(theta), sin(theta)], [0, -sin(theta), cos(theta)]])
```

Comment:
I'm not getting the same direction cosine matrix. Here is my solution to this problem:

```
In [70]: N = me.ReferenceFrame('N')

In [71]: A = me.ReferenceFrame('A')

In [72]: As = me.ReferenceFrame('As')

In [73]: As.orient_axis(N, N.z, sm.pi/2)

In [74]: A.orient_axis(As, N.x, theta)

In [75]: A.dcm(N)
Out[75]: 
Matrix([
[ 0,  cos(theta(t)), sin(theta(t))],
[-1,              0,             0],
[ 0, -sin(theta(t)), cos(theta(t))]])

In [76]: N.x.express(A)
Out[76]: - A.y

In [78]: A.ang_vel_in(N)
Out[78]: - Derivative(theta(t), t)*As.y

In [79]: A.ang_vel_in(N).express(N)
Out[79]: Derivative(theta(t), t)*N.x
```

Reply:
Here is my sketch:
![image](https://user-images.githubusercontent.com/276007/123751123-a8f25280-d8b7-11eb-9946-b509e37daba8.png)

Reply:
> I'm not getting the same direction cosine matrix. Here is my solution to this problem:
> 
> ```
> In [70]: N = me.ReferenceFrame('N')
> 
> In [71]: A = me.ReferenceFrame('A')
> 
> In [72]: As = me.ReferenceFrame('As')
> 
> In [73]: As.orient_axis(N, N.z, sm.pi/2)
> 
> In [74]: A.orient_axis(As, N.x, theta)
> In [76]: N.x.express(A)
> Out[76]: - A.y
> 

> ```

`A.x` should be aligned to `N.x` . You are getting `N.x`  aligned to `A.y`

Reply:
Here is the corrected code for the figure:

```
In [89]: N = me.ReferenceFrame('N')

In [90]: A = me.ReferenceFrame('A')

In [91]: As = me.ReferenceFrame('As')

In [92]: As.orient_axis(N, N.z, sm.pi)

In [93]: A.orient_axis(As, N.x, theta)

In [94]: A.dcm(N)
Out[94]: 
Matrix([
[-1,              0,              0],
[ 0, -cos(theta(t)), -sin(theta(t))],
[ 0, -sin(theta(t)),  cos(theta(t))]])

In [95]: N.x.express(A)
Out[95]: - A.x

```

Reply:
Same result:

```
In [96]: N = me.ReferenceFrame('N')

In [97]: A = me.ReferenceFrame('A')

In [98]: As = me.ReferenceFrame('As')

In [99]: As.orient_axis(N, -N.z, sm.pi)

In [100]: A.orient_axis(As, N.x, theta)

In [101]: A.dcm(N)
Out[101]: 
Matrix([
[-1,              0,              0],
[ 0, -cos(theta(t)), -sin(theta(t))],
[ 0, -sin(theta(t)),  cos(theta(t))]])
```